### PR TITLE
Format Vue SFC root blocks

### LINKED ISSUE #6298: Pug plugin doesn't work for the .vue files
**Environments:**
- Prettier Version: 1.18.2
- Running Prettier via: CLI
- Runtime: node.js v8.15.0
- Operating System: MacOS

**Steps to reproduce:**
- clone https://github.com/xorik/prettier-pug-vue
- run `prettier --write *.vue`

**Expected behavior:**
Template inside the `pug.vue` should be prettified

**Actual behavior:**
template inside the `pug.vue` isn't touched by prettier

If you run `prettier --write *.pug`, you can see that the plugin itself works perfectly.
The plugin developers suggested to write here: (https://github.com/prettier/plugin-pug/issues/5)

### PR DESCRIPTION
**Pull request by fisker**
To play with it,

```console
mkdir prettier-vue-pug-test
cd prettier-vue-pug-test
yarn init -y
yarn add fisker/prettier#vue-embed @prettier/plugin-pug
code test.vue
npx prettier test.vue --plugin-search-dir=.
```

`--plugin-search-dir` won't need when we fix https://github.com/prettier/prettier/issues/8474

Fixes #6298

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by Shinigami92** [comment]
First local test seems to work

image

OS: Windows 10
Used Terminal: cmd

---

Blank lines above and below the code could be on my side
(especially the blank line below)

**Comment by Shinigami92** [comment]
With logs enabled and using `yarn link "@prettier/plugin-pug"`:

```shell
D:\Shinigami\Documents\Projects\prettier-vue-pug-test>npx prettier test.vue --plugin-search-dir=.
test.vue[parsers:pug:preprocess]: { text: '\n  div.text( color =   "primary",  disabled    )\n' }
[parsers:pug:parse]: { text: '\n  div.text( color =   "primary",  disabled    )\n' }
[printers:pug-ast:embed]: {
  "stack": [
    [
      {
        "type": "indent",
        "loc": {
          "start": {
            "line": 2,
            "column": 1
          },
          "end": {
            "line": 2,
            "column": 3
          }
        },
        "val": 2
      },
      {
        "type": "tag",
        "loc": {
          "start": {
            "line": 2,
            "column": 3
          },
          "end": {
            "line": 2,
            "column": 6
          }
        },
        "val": "div"
      },
      {
        "type": "class",
        "loc": {
          "start": {
            "line": 2,
            "column": 6
          },
          "end": {
            "line": 2,
            "column": 11
          }
        },
        "val": "text"
      },
      {
        "type": "start-attributes",
        "loc": {
          "start": {
            "line": 2,
            "column": 11
          },
          "end": {
            "line": 2,
            "column": 12
          }
        }
      },
      {
        "type": "attribute",
        "loc": {
          "start": {
            "line": 2,
            "column": 13
          },
          "end": {
            "line": 2,
            "column": 32
          }
        },
        "name": "color",
        "val": "\"primary\"",
        "mustEscape": true
      },
      {
        "type": "attribute",
        "loc": {
          "start": {
            "line": 2,
            "column": 35
          },
          "end": {
            "line": 2,
            "column": 43
          }
        },
        "name": "disabled",
        "val": true,
        "mustEscape": true
      },
      {
        "type": "end-attributes",
        "loc": {
          "start": {
            "line": 2,
            "column": 47
          },
          "end": {
            "line": 2,
            "column": 48
          }
        }
      },
      {
        "type": "outdent",
        "loc": {
          "start": {
            "line": 3,
            "column": 1
          },
          "end": {
            "line": 3,
            "column": 1
          }
        }
      },
      {
        "type": "eos",
        "loc": {
          "start": {
            "line": 3,
            "column": 1
          },
          "end": {
            "line": 3,
            "column": 1
          }
        }
      }
    ]
  ]
}
[printers:pug-ast:print]:
  .text(color="primary", disabled)

<template lang="pug">

  .text(color="primary", disabled)

</template>

<script>
export default {
  name: "Hello",
};
</script>

```

Thats the **first time** my plugin receiving something from vue `<template lang="pug">` 🎉

**Comment by fisker** [comment]
I'm not sure about that, `vue-indent-script-and-style` mean to indent `<script>` and `<style>`, should we apply to those blocks ? They are not `<script>` :smile:

There might be another problem(but I'm not sure), plugins will format these blocks, if this language/plugin indent sensitive? Maybe python?

File: tests_config/run_spec.js [hunk 1]
```diff
@@ -56,8 +56,11 @@ const isTestDirectory = (dirname, name) =>
   dirname.startsWith(path.join(__dirname, "../tests", name));
 
 global.run_spec = (fixtures, parsers, options) => {
-  // Can't test plugins on standalone
+  // Can't test plugins on standalone, a empty test to pass
```
Comment (thorn0) [inline tests_config/run_spec.js]: You can test them if you pass them as objects, not strings.

**Comment by Shinigami92** [comment]
> I'm not sure about that, `vue-indent-script-and-style` mean to indent `<script>` and `<style>`, should we apply to those blocks ? They are not `<script>` 😄
> 
> There might be another problem(but I'm not sure), plugins will format these blocks, what if this language/plugin indent sensitive? Maybe `python`? or does `pug` care?

In any case, it makes sense for HTML or XML-based languages
```
<template>
  <div>
    <p>some content</p>
  </div>
</template>
```

But for non XML based languages like `pug`? 🤷‍♂️
```
<template lang="pug">
div
  p some content
</template>
```

IMO makes no sense for such languages

File: src/language-html/printer-html.js [hunk 2]
```diff
@@ -63,7 +64,42 @@ function concat(parts) {
...
+
+      if (isVueNonHtmlBlock(node, options)) {
+        const parser = inferScriptParser(node, options);
+        if (!parser) {
+          return;
+        }
+
+        let doc;
+        try {
+          doc = textToDoc(getNodeContent(node, options), { parser });
```
Comment (fisker) [inline src/language-html/printer-html.js:83]: This text pass to parser need discuss too, 

```js
<custom lang="pug">
  div
</custom>
```

What we should pass to plugin? 
- `"\n  div\n"` (current)
- `"  div"` (trim leading and trailing empty lines)
- `"  div\n"` (trim ltrailing empty lines)
- ~`"div"` (this should not consider)~

File: tests_config/run_spec.js (hunk 1 above)
Comment (fisker) [reply tests_config/run_spec.js]: I've tried this before... I forgot. Can't work, maybe because there are `function`s?

Comment (thorn0) [reply tests_config/run_spec.js]: Okay. Let's keep it this way. I might look at this later.

Comment (fisker) [reply tests_config/run_spec.js]: Are you talking about `a empty test to pass`? This not work too, snapshots don't match.

I ignored it in `jest.config.js` for now.

Comment (thorn0) [reply tests_config/run_spec.js]: No, about passing the plugin as an object. It should work. If it doesn't, something isn't right.

File: src/language-html/printer-html.js (hunk 2 above)
Comment (Shinigami92) [reply src/language-html/printer-html.js:83]: I would prefer the exact behavior as currently with the embedded markdown

Comment (fisker) [reply src/language-html/printer-html.js:83]: Which is?

if same

````
```pug
  div
```
````

Comment (fisker) [reply src/language-html/printer-html.js:83]: Seems removed one line on each side, means extra blank lines kept. https://github.com/prettier/prettier/blob/master/src/language-markdown/utils.js#L220

Comment (Shinigami92) [reply src/language-html/printer-html.js:83]: _Beside the fact that my plugin has a bug_

_embedded markdown_

````
```pug
  div
```
````

**currently** formats to (bug)

````
```pug
| div
```
````

but **should** format to 

````
```pug
div
```
````

**Comment by Shinigami92** [comment]
`lang="uppercase-rocks"` 😂
I like that 👍

**Comment by fisker** [comment]
> `lang="uppercase-rocks"` 😂
> I like that 👍

NO NO NO, YOU LIKE THAT!

**Comment by Shinigami92** [comment]
> > `lang="uppercase-rocks"` 😂
> > I like that 👍
> 
> NO NO NO, YOU LIKE THAT!

OH MY GOD, IT TOOK ME SO LONG TO UNDERSTAND THIS JOKE XD

**Comment by fisker** [comment]
If no objection about `vue-indent-script-and-style`, approve? 

@prettier/core

**Comment by thorn0** [comment]
Changelog?

**Comment by fisker** [comment]
Shared a changelog with #8023

**Comment by fisker** [comment]
I forgot discuss this with you guys, now we have extra `\n` for `pug`

```html
<template lang="pug">
.text(color="primary", disabled="true")

</template>
```

First, plugins prints own `\n`? Why not add this in `core.format()`? @thorn0 Do you know the reason?

With builtin embed languages, we use `stripTrailingHardline` to remove the `\n`, but `@prettier/plugin-pug` returns a string. Should we keep it? or remove it? or ask plugins (in this case @prettier/plugin-pug) to remove it?

@thorn0 @Shinigami92 @sosukesuzuki Thoughts?

**Comment by Shinigami92** [comment]
I think this can be managed on my (plugin) side. So when the [embed](https://github.com/prettier/plugin-pug/blob/67cb1ebbe8f2f94e8a92300782080e5d8b5abc9d/src/index.ts#L86) function is called, I could remove it at the end.

But why is `markdown` already working? 🤔

**Comment by fisker** [comment]
interesting, I had a quick look, seems `markdown` did not insert a `hardline` after the code, it use the `\n` return from `textToDoc`, https://github.com/prettier/prettier/blob/e54d4fc01af5957d1ba147ffed72417055fd8daa/src/language-markdown/embed.js#L36

But, I don't think that's the right way, if a plugin didn't add `final_newline`, it will be 

````markdown
```foo
text```
````
?

**Comment by thorn0** [comment]
@fisker It's unlikely there is a good reason for that. Just history I guess. Overall, 3rd party plugins should behave the same way the builtin plugins do, and we need to document these requirements. However, in this case, we need to decide first if we want to change this behavior for the builtin plugins and move the addition of the final new line to the core. I think this change would make sense.

**Comment by fisker** [comment]
I'm going to remove this `\n` manually if it's a string, and move this discussion to a new issue. Let's decide there, sounds good?

**Comment by thorn0** [comment]
@fisker Good

**Comment by fisker** [comment]
#8582

**Comment by fisker** [comment]
@thorn0 You might want look into this https://github.com/prettier/prettier/pull/8465#discussion_r435086225

**Pull request merged**